File: src/tools/gem5log.php
#!/usr/bin/env php
<?php
function usage() {
    die("Usage: (trace|functime) <binary>...\n"
        . "With the gem5 log in stdin, which should have Exec and ExecPC enabled.\n");
}

function bchexdec($hex) {
    if(strlen($hex) == 1)
        return hexdec($hex);
    else {
        $remain = substr($hex, 0, -1);
        $last = substr($hex, -1);
        return bcadd(bcmul(16, bchexdec($remain)), hexdec($last));
    }
}

function addr2func($addr) {
    global $funcs;
    static $last = null;
    if($last !== null && bccomp($addr,$last[0]) >= 0 && bccomp($addr,$last[0] + $last[1]) <= 0)
        return $last;

    foreach($funcs as $f) {
        if(bccomp($addr,$f[0]) >= 0 && bccomp($addr,$f[0] + $f[1]) <= 0) {
            $last = $f;
            return $f;
        }
    }
    return false;
}

function funcname($addr) {
    $f = addr2func($addr);
    if($f !== false) {
        return sprintf("\033[1m%s\033[0m @ %s+0x%x",
            basename($f[3]),strtok($f[2],"("),bcsub($addr,$f[0]));
    }
    return sprintf("<Unknown>: %08x", $addr);
}

function func_compare($a, $b) {
    return $a[0] < $b[0];
}

if($argc < 2)
    usage();
$mode = $argv[1];
if($mode != 'trace' && $mode != 'functime')
    usage();

$funcs = array();
for($i = 2; $i < $argc; $i++) {
    $f = popen("nm -SC " . escapeshellcmd($argv[$i]), "r");
    if(!$f)
        die("Unable to open " . $argv[$i]);
    while(($line = fgets($f)) !== false) {
        if(preg_match('/^([a-f0-9]+) ([a-f0-9]+) (t|T) ([A-Za-z0-9_:\.\~ ]+)/', $line, $match))
            $funcs[] = array(bchexdec($match[1]), bchexdec($match[2]), $match[4], $argv[$i]);
    }
    pclose($f);
}

usort($funcs,'func_compare');

$f = @fopen("php://stdin", "r");
if(!$f)
    die("Unable to open stdin");

if($mode == 'trace') {
    while(($line = fgets($f)) !== false) {
        if(preg_match('/(\d+): (pe\d+\.cpu) (\S+) : 0x([a-f0-9]+)\s*@\s*.+?\s*:\s*(.*)$/', $line, $match)) {
            $func = funcname(bchexdec($match[4]));
            printf("%7s: %s: %s : %s\n", $match[1], $match[2], $func, $match[5]);
        }
        else
            echo $line;
    }
}
else {
    $pes = array();
    while(($line = fgets($f)) !== false) {
        if(preg_match('/(\d+): (pe(\d+)\.cpu) (\S+) : 0x([a-f0-9]+)/', $line, $match)) {
            $addr = bchexdec($match[5]);
            $func = addr2func($addr);
            if($func === false)
                $func = array($addr, $addr, sprintf('%08x', $addr), '<Unknown>');
            $pe = $match[3];

            if(!isset($pes[$pe]) || $pes[$pe][0][2] != $func[2]) {
                if(isset($pes[$pe])) {
                    printf("%9s: %s: [%9u ticks] \033[1m%10.10s\033[0m @ %s+0x%x\n",
                        $match[1], $match[2], bcsub($match[1], $pes[$pe][1]),
                        basename($pes[$pe][0][3]), $pes[$pe][0][2],
                        bcsub($pes[$pe][2], $pes[$pe][0][0]));
                }
                $pes[$pe] = array($func, $match[1], $addr);
            }
        }
        else if(preg_match('/^(\d+): (.*)$/', $line, $match))
            printf("%9s: %s\n", $match[1], $match[2]);
    }
}

if(!feof($f))
    die("Something went wrong");
fclose($f);
?>

File: src/tools/gem5log.py
#!/usr/bin/env python

import sys
import subprocess
import re
from os.path import basename

funcs = []

def usage():
    print("Usage: %s (trace|functime) <binary>..." % sys.argv[0], file=sys.stderr)
    print("  With run/gem5.log as stdin and Exec,ExecPC enabled.", file=sys.stderr)
    sys.exit(1)

def addr2func(funcs, addr):
    if addr2func.last:
        if addr >= addr2func.last['addr'] and addr < addr2func.last['addr'] + addr2func.last['size']:
            return addr2func.last

    for f in funcs:
        if addr >= f['addr'] and addr < f['addr'] + f['size']:
            addr2func.last = f
            return f
    return {}

addr2func.last = {}

def funcname(funcs, addr):
    f = addr2func(funcs, addr)
    if not f:
        return '<Unknown>: %x' % addr

    name = f['name'].decode('ASCII')
    return '\033[1m%s\033[0m @ %s+0x%x' % (basename(f['bin']), name, addr - f['addr'])

if len(sys.argv) < 2:
    usage()

# read symbols
for i in range(2, len(sys.argv)):
    proc = subprocess.Popen(['nm', '-SC', sys.argv[i]], stdout=subprocess.PIPE)
    while True:
        line = proc.stdout.readline()
        if not line:
            break

        m = re.match(b'^([a-f0-9]+) ([a-f0-9]+) (t|T|w|W) ([A-Za-z0-9_:\.\~<>, ]+)', line)
        if m:
            funcs.append({
                'addr': int(m.group(1), 16),
                'size': int(m.group(2), 16),
                'name': m.group(4),
                'bin': sys.argv[i]
            })

# sort symbols by address
funcs = sorted(funcs, key=lambda f: f['addr'])

# read log lines
if sys.argv[1] == 'trace':
    while True:
        line = sys.stdin.readline()
        if not line:
            break

        m = re.match('^\s*(\d+): (pe\d+\.cpu\d*) (\S+) : 0x([a-f0-9]+)\s*(?:@\s+)?[^@]+?\s+:\s+(.*)$', line)
        if m:
            func = funcname(funcs, int(m.group(4), 16));
            print("%7s: %s: %s : %s" % (m.group(1), m.group(2), func, m.group(5)));
        else:
            print(line.rstrip())
elif sys.argv[1] == 'functime':
    pes = {}
    while True:
        line = sys.stdin.readline()
        if not line:
            break

        m = re.match('^\s*(\d+): (pe(\d+)\.cpu\d*) (\S+) : 0x([a-f0-9]+).*', line)
        if m:
            # find function
            addr = int(m.group(5), 16)
            func = addr2func(funcs, addr)
            if not func:
                func = {
                    'addr': addr,
                    'name': b'<Unknown>: %x' % addr,
                    'bin': '<Unknown>'
                }

            pe = int(m.group(3))

            # is it a new function?
            if not pe in pes or pes[pe]['func']['name'] != func['name']:
                tick = int(m.group(1))
                if pe in pes:
                    # print ticks for last function
                    name = pes[pe]['func']['name'].decode('ASCII')
                    print("%9s: %s: [%9u ticks] \033[1m%10.10s\033[0m @ %s+0x%x" % (
                        m.group(1), m.group(2), tick - pes[pe]['tick'],
                        basename(pes[pe]['func']['bin']), name,
                        pes[pe]['addr'] - pes[pe]['func']['addr']
                    ))
                # remember function start
                pes[pe] = {'func': func, 'tick': tick, 'addr': addr}
        else:
            m = re.match('^\s*(\d+): (.*)$', line)
            if m:
                print("%9s: %s" % (m.group(1), m.group(2)))
else:
    usage()
